### DISCUSSION TITLE
Crash when navigating back from navigationlink in sheet/fullscreenCover (iPhone 12 pro max)

### DISCUSSION THREAD

**Discussion by Nicolaidam** [pointfreeco/swift-composable-architecture#1417] [General]
### Description

I'm experiencing a weird crash when navigating back programmatically from a navigationlink flow inside a sheet/fullscreenCover. 

I have only been able to reproduce the crash using iPhone 12 pro max with iOS 16.0.2. On other devices such as iPhone 13 pro, iPhone 11 and iPhone 8 plus i have not been able to reproduce, even when using 16.0.2. I don't know it is just an iOS 16 thing or if other versions of iOS are also affected. (or if it a bug in the actual device or whatever)

This line is printed before the crashes and points that an action is called from a nil state. But as i understand it is this hard to prevent since the Binding set method is called after the view is set to nil. This message is just  a warning and should not crash the app, right?

> 2022-09-27 14:38:38.691705+0100 NavigationCrash[1339:155431] [ComposableArchitecture] An "optional" reducer at "NavigationCrash/NavigationCrashApp.swift:37" received an action when state was "nil". …
> 
>   Action:
>     FullscreenCoverAction.navigationChanged
> 
> This is generally considered an application logic error, and can happen for a few reasons:
> 
> • The optional reducer was combined with or run from another reducer that set "FullscreenCoverState" to "nil" before the optional reducer ran. Combine or run optional reducers before reducers that can set their state to "nil". This ensures that optional reducers can handle their actions while their state is still non-"nil".
> 
> • An in-flight effect emitted this action while state was "nil". While it may be perfectly reasonable to ignore this action, you may want to cancel the associated effect before state is set to "nil", especially if it is a long-living effect.
> 
> • This action was sent to the store while state was "nil". Make sure that actions for this reducer can only be sent to a view store when state is non-"nil". In SwiftUI applications, use "IfLetStore".
> (lldb) 

I can't guarantee that my implementation is correct, so please let me know if there is something I'm missing.

### Checklist

- [X] If possible, I've reproduced the issue using the `main` branch of this package.
- [X] This issue hasn't been addressed in an [existing GitHub issue](https://github.com/pointfreeco/swift-composable-architecture/issues).

### Expected behavior

I expect that I'm not crashing when navigating back from navigationLink flow in sheet/fullscreenCover using iPhone 12 pro max.

### Actual behavior

Crash: Thread 1: EXC_BREAKPOINT (code=1, subcode=0x1d13a4208)

![image](https://user-images.githubusercontent.com/32563707/192528773-3fa3814e-6884-4003-8848-59645ce09023.png)

### Steps to reproduce

https://user-images.githubusercontent.com/32563707/192541045-c6f285c8-e566-4a7d-962d-1ab6ae2e4a58.MP4

[NavigationCrash.zip](https://github.com/pointfreeco/swift-composable-architecture/files/9656646/NavigationCrash.zip)

1. Build attached project
2. Run the app on an iPhone 12 pro max with iOS 16.0.2
3. Tap on "Show Fullscreen"
4. Tap on "Navigate To NavigationLink"
5. Tap on "Close fullscreen cover"

### The Composable Architecture version information

0.40.2

### Destination operating system

iOS 16.0.2

### Xcode version information

Version 14.0.1 (14A400), also tried 14.1 beta

**Comment by mbrandonw** [comment]
I wasn't able to reproduce the crash with just simulators, and I don't have a device on exactly 16.0.2, but I feel pretty certain this is a SwiftUI bug and not a problem with the library (especially since it seems to be SDK/device specific). I would recommend recreating the app in vanilla SwiftUI and see if it behaves the same, and file a feedback if it does. The warning you are seeing is just a warning, and does not cause crashes.

There are still all types of crashes lurking in SwiftUI's navigation APIs. For example, if you have a navigation stack with A->B->C on the stack, and B has state for a sheet, then popping back C->B causes a crash. So, I wouldn't be surprised if dismissing a sheet with a drilled down navigation causes problems.

I'm going to convert this to a discussion until we can pinpoint that this is definitely an issue with the library.

**Reply by Nicolaidam** [reply]
Ahh okay, thank you @mbrandonw

You're probably right it is a SwiftUI bug.. but it's so weird I can only reproduce it on this damn iPhone 12 pro max... 🫠

I will try to reproduce it in vanilla SwiftUI. 🙏

**Reply by Nicolaidam** [reply]
[NavigationCrash 2.zip](https://github.com/pointfreeco/swift-composable-architecture/files/9657458/NavigationCrash.2.zip)

I have now worked on reimplementing the app in vanilla SwiftUI/Combie. Unfortunately I'm not able to reproduce a crash. However I'm not completely sure that I have recreated the exact same behaviour as in the tca example so i would love to get your feedback.

**Reply by tgrapperon** [reply]
From a quick look, there is a very slight difference because your embedded models that you host as `@Published` will only emit in the app model when you change their instance, not their internal value.
One thing you can do in your original app is to narrow each `ViewStore` to only the bare minimum required to display the view. For example, in `AppView`, you can try to observe only `\.fullscreen` (`WithViewStore(store, observe: \.fullscreen) {…}`, and adapt the get part of the binding to `$0` instead of `$0.fullscreen`). SwiftUI navigation can behave erratically when its `@ObservedObject`s (here `ViewStore`s) are emitting in positions where their view is only presenting. It is possible that what is usually an erratic behaviors becomes a crash on iPhone 12 pro max (these larger phones adapt to the `.regular` horizontal size class & presentation in landscape, so they likely activate different code-paths than smaller phones when presenting/navigating). Same with `secondView` for the  `FullScreenCover` domain. This would also makes the TCA version much closer to your vanilla one.

**Reply by Nicolaidam** [reply]
Thanks for taking your time @tgrapperon 

I have now added the _WithViewStore(store, observe: \.fullscreen)_ to all the viewstores, and I'm still experiencing the same behaviours. 

Any ideas on how I can change the vanilla swiftui version to get even closer to the tca version? I'm not really sure how to listen for the close action in the vanilla swiftui without changing a @Published state value. That is the biggest difference as I understand it.

**Reply by Nicolaidam** [reply]
[NavigationCrash 3.zip](https://github.com/pointfreeco/swift-composable-architecture/files/9658172/NavigationCrash.3.zip)

**Reply by tgrapperon** [reply]
I've looked closer and putting your `NavigationLink` into a `.background`, furthermore with an `EmptyView` label is a little unconventional. Does it still crash if you put it beside the button in a `VStack`, with a `Text` label for example?

**Reply by Nicolaidam** [reply]
Thanks, I just tried to put the `NavigationLink` in the `VStack` instead of `.background`, and changed the `EmptyView` to a `Text` instead. But still crashing when closing the sheet.
<img width="589" alt="image" src="https://user-images.githubusercontent.com/32563707/192589756-49db48f6-fab9-4aa5-b460-c8f8727cef6a.png">

**Reply by erikolsson** [reply]
I wonder if this is the same issue I was seeing: https://github.com/pointfreeco/swift-composable-architecture/discussions/1379#discussioncomment-3706247

**Reply by tgrapperon** [reply]
It was a long shot. I would then work backward to remove TCA from the example, checking at each step if it fixes the problem. For example, you can replace this whole `IfLetStore` by a `Button` that sends `.fullscreenSheetAction(.buttonTapped)`, and change `SecondViewState` by a simple `Bool`, then I would unconstrain the `isActive` binding of this `NavigationLink`, and so on. The idea is to narrow the problem until it doesn't crash, so we can have a better understanding at what's happening. Then, if we can reverse this last step as a vanilla SwiftUI implementation and get the crash back, it would likely confirms that the issue is pure SwiftUI (which is already very likely). If this event never happens, you will end up with a vanilla SwiftUI app that crashes, and that will be a definitive confirmation.
You can send `self.objectWillChange()` from an `ObservableObject` if you want to handle invalidation yourself instead of using a `@Published` property. You ideally need to do it before assigning a new value.

**Reply by tgrapperon** [reply]
@erikolsson @Nicolaidam Indeed, I would also try to permute `WithViewStore` and `NavigationView` in any case.

**Reply by Nicolaidam** [reply]
[NavigationCrash4.zip](https://github.com/pointfreeco/swift-composable-architecture/files/9683235/NavigationCrash4.zip)

Thanks for the inputs. I have tried to decouple as much code as i can from TCA. However I'm still using one IfLetStore as I'm not sure how to remove more of TCA while having the exact same functionality. Im still able to reproduce the bug. Not only on  iPhone 12 pro max (iOS 16.0.2) but also iPhone 12 mini (iOS 16.0.2) and iPhone 13 pro (iOS 16.0.0). I have updated my own iPhone 13 pro to 16.1 beta and I can't reproduce the issue on that device anymore. So maybe it will be fixed in the 16.1 update.

You can also find the vanilla implementation in the zip that is not having any issues when navigating back. However I'm not completely sure that it reproduce the exact same scenario.

**Reply by Nicolaidam** [reply]
This is the crash report I get from Crashlytics, maybe it makes sense for any of you? 🤔
<img width="1254" alt="image" src="https://user-images.githubusercontent.com/32563707/193250586-3c55ab28-aee9-434e-a2ab-7025b7c7be66.png">

**Reply by tgrapperon** [reply]
Thanks for the feedback @Nicolaidam. It looks like the system is trying to retain a parent `UIViewController` that was already released. This thing happens on Apple's side and there is probably very little we can do to avoid tripping on it.

One thing that could be slightly odd for the system is the `NavigationLink` activating because of an user action originating from elsewhere. It would be interesting to check if it still crashes when interacting directly with the NavigationLink, and if so, when removing the `isActive` part and let it unconstrained. If it definitely looks like a SwiftUI/UIKit bug, it would be good to understand what can lead to this behavior.

Assuming it would still crash, you can remove the `IfLetStore` in two steps: first, you can scope regularly onto an optional `FullscreenState?` while providing a default value: `store.scope(state: { $0.fullscreen ?? .init(), action: …)` and directly create a `FullScreenCoverView` with this store. After checking that it still crashes, you can then convert this view into a `View` that accepts the two actions as closure parameters (as at this stage, you don't have a `navigateToLastView` anymore and it is still crashing.
It it still crashes after that, you can replicate the app's `ViewStore` as an `ObservableObject` (I think this is would be the last bit of TCA).

**Reply by Nicolaidam** [reply]
Thank you for the quick reply @tgrapperon. You definitely have some great points!

I just tried to remove the NavigationLink isActive binding and use the simpel NavigationLink initialiser instead. Unfortunately I can still replicate the issue.

struct FullscreenCover: View {
    
    var store: Store<FullscreenCoverState, FullscreenCoverAction>
    
    init(store: Store<FullscreenCoverState, FullscreenCoverAction>) {
        self.store = store
    }
    
    var body: some View {
        WithViewStore(store) { viewStore in
            NavigationView {
                VStack {
                    NavigationLink {
                        Button("close") {
                            viewStore.send(.closeSheetTapped)
                        }
                    } label: {
                        Text("Navigate to NavigationLink")
                    }
                }
            }
        }
    }

Afterwards I tried to remove the IfLetStore as you suggested and after doing that I'm not able to reproduce the crash anymore!

<img width="991" alt="image" src="https://user-images.githubusercontent.com/32563707/193260269-916c0c8b-a817-4960-b4a7-d273caf70ade.png">


Could this mean that the issue potentially is in the IfLetStore or what can we conclude from this?

**Reply by Nicolaidam** [reply]
[NavigationCrash5.zip](https://github.com/pointfreeco/swift-composable-architecture/files/9683733/NavigationCrash5.zip)

**Reply by tgrapperon** [reply]
This is odd because `IfLetStore` implementation is very straightforward. Would you mind checking what happens if you copy/paste the following and replace `IfLetStore` by `IfLetStore2` in your implementation? It simply avoids to perform a trick to build an internal type. This implementation defines its own type to play the role of this internal type. While doing so, it also builds `EmptyView` explicitly instead of a `nil` View.
```swift
public enum Either<First, Second> {
  case first(First)
  case second(Second)
}

extension Either: View where First: View, Second: View {
  public var body: some View {
    switch self {
    case .first(let first):
      first
    case .second(let second):
      second
    }
  }
}

public struct IfLetStore2<State, Action, Content: View>: View {
  private let content: (ViewStore<State?, Action>) -> Content
  private let store: Store<State?, Action>

  public init<IfContent, ElseContent>(
    _ store: Store<State?, Action>,
    @ViewBuilder then ifContent: @escaping (Store<State, Action>) -> IfContent,
    @ViewBuilder else elseContent: @escaping () -> ElseContent
  ) where Content == Either<IfContent, ElseContent> {
    self.store = store
    self.content = { viewStore in
      if var state = viewStore.state {
        return .first(ifContent(
          store.scope {
            state = $0 ?? state
            return state
          }
        ))
      } else {
        return .second(elseContent())
      }
    }
  }
  
  public init<IfContent>(
    _ store: Store<State?, Action>,
    @ViewBuilder then ifContent: @escaping (Store<State, Action>) -> IfContent
  ) where Content == Either<IfContent, EmptyView> {
    self.store = store
    self.content = { viewStore in
      if var state = viewStore.state {
        return .first(ifContent(
            store.scope {
              state = $0 ?? state
              return state
            }
          )
        )
      } else {
        return .second(EmptyView())
      }
    }
  }

  public var body: some View {
    WithViewStore(
      self.store,
      observe: { $0 },
      removeDuplicates: { ($0 != nil) == ($1 != nil) },
      content: self.content
    )
  }
}
```

**Reply by Nicolaidam** [reply]
cool thank you man. It is still crashing when I use that IfLetStore2 implementation.

**Reply by tgrapperon** [reply]
Thanks. The only thing I'm seeing to be slightly off is the closure being doubly escaped. We can probably simplify the implementation to do it only once in `WithViewStore`, or even forgo `WithViewStore` at the price of one additional generic. Here are two `IfLetStore3` and `IfLetStore4` replacements. If it still crashes for both, I can't really see what's triggering the issue.

```swift
public struct IfLetStore3<State, Action, Content: View>: View {
  private let content: Content
  public init<IfContent, ElseContent>(
    _ store: Store<State?, Action>,
    @ViewBuilder then ifContent: @escaping (Store<State, Action>) -> IfContent,
    @ViewBuilder else elseContent: () -> ElseContent
  ) where Content == WithViewStore<State?, Action, _ConditionalContent<IfContent, ElseContent>> {
    let elseContent = elseContent()
    self.content = WithViewStore<State?, Action, _ConditionalContent<IfContent, ElseContent>>(
      store,
      observe: { $0 },
      removeDuplicates: { ($0 != nil) == ($1 != nil) },
      content: {
        if var state = $0.state {
          ifContent(
            store.scope {
              state = $0 ?? state
              return state
            }
          )
        } else {
          elseContent
        }
      }
    )
  }

  public init<IfContent>(
    _ store: Store<State?, Action>,
    @ViewBuilder then ifContent: @escaping (Store<State, Action>) -> IfContent
  ) where Content == WithViewStore<State?, Action, IfContent?> {
    self.content = WithViewStore<State?, Action, IfContent?>(
      store,
      observe: { $0 },
      removeDuplicates: { ($0 != nil) == ($1 != nil) },
      content: {
        if var state = $0.state {
          ifContent(
            store.scope {
              state = $0 ?? state
              return state
            }
          )
        }
      }
    )
  }
  public var body: some View {
    content
  }
}

public struct IfLetStore4<State, Action, IfContent: View, ElseContent: View>: View {
  private let store: Store<State?, Action>
  @ObservedObject private var viewStore: ViewStore<State?, Action>
  private let ifContent: (Store<State, Action>) -> IfContent
  private let elseContent: ElseContent

  public init(
    _ store: Store<State?, Action>,
    @ViewBuilder then ifContent: @escaping (Store<State, Action>) -> IfContent,
    @ViewBuilder else elseContent: () -> ElseContent
  ) where ElseContent == EmptyView?  {
    self.store = store
    self.viewStore = ViewStore(store, removeDuplicates: { ($0 != nil) == ($1 != nil) })
    self.ifContent = ifContent
    self.elseContent = elseContent()
  }
  
  public init(
    _ store: Store<State?, Action>,
    @ViewBuilder then ifContent: @escaping (Store<State, Action>) -> IfContent
  ) where ElseContent == EmptyView?  {
    self.store = store
    self.viewStore = ViewStore(store, removeDuplicates: { ($0 != nil) == ($1 != nil) })
    self.ifContent = ifContent
    self.elseContent = nil
  }

  public var body: some View {
    if var state = viewStore.state {
      ifContent(
        store.scope {
          state = $0 ?? state
          return state
        }
      )
    } else {
      elseContent
    }
  }
}
```

**Reply by Nicolaidam** [reply]
Good point. I have now tried all the different IfLetStore implementations on a iPhone 12 pro max and still crashing on all of the versions.

**Reply by tgrapperon** [reply]
It's definitely an Apple bug, but I can't see how to avoid triggering it when using `IfLetStore`. Maybe you can use the scope with fallback to a default state value instead of `IfLetStore` as a workaround in your app. It shouldn't cause any issue as this screen is also controlled by the `.fullscreenCover` binding.

**Reply by Nicolaidam** [reply]
Sorry for the delay, I totally forgot to respond, since i was very busy with a new release.
As you suggested the problem was fixed by doing a scope fallback to a default state, instead of the IfLetStore. This seems to be working and we he haven't had the crash since that implementation change. 

So I just wanna thank you so much for help @tgrapperon 🙏 This community is awesome!

**Comment by artem-krachulov-paysera** [comment]
@Nicolaidam Have the same issue, when trying to dismiss **sheet** with navigation stack (2 screens) by setting sheet **state** to **nil**. Reproducible only on device (sim fine) and probably on iOS 16.0+. 

**How to reproduce:**
1. Present **sheet** with **NavigationView**
2. Push the next screen by **NavigationLink**
3. Dismiss **sheet** by setting **state.sheetState = nil**

The crash is not consistent.

<img width="485" alt="Screenshot 2022-11-04 at 10 03 19 AM" src="https://user-images.githubusercontent.com/7997760/199922949-3f45c0eb-fe5c-41c3-81f7-d493e89db982.png">
